### ISSUE TITLE
Inode.Val.t is not always a root

### ISSUE THREAD

**Issue by Ngoguey42** [mirage/irmin#1266]
TLDR; there is a way to violate the unique representation invariant, I doubt anything is broken anywhere today.

---

In Inode we make the assumption that the `Val_impl.t` directly pointed by a `Val.t` is always the root of an inode tree, which is meaningful because such a `Val_impl.t` is the only kind of `Val_impl.t` that goes through the `stabilize` function.

The problem is that `Inode.find : _ t -> hash -> _` may be called using a hash of an intermediate inode.

---

Here is a scenario that produces such a call to find:
- We first call `pred_node` (https://github.com/mirage/irmin/blob/master/src/irmin-pack/layered/irmin_pack_layered.ml#L570) using the hash `h0` of the root of an inode tree
    - It may yield the hash `h1` from an intermediate inode
- `pred_node` is then called again but using `h1` this time
    - `v` is now a `Inode.Val.t` that doesn't point to the root of an inode tree.

If `v` has `Conf.entries < #entries <= Conf.stable_hash`, we have an unstable `Val.t` where we would normally expect a stable one.

If you add a value to that inode using `Inode.Val.add` and then remove it using `Inode.Val.remove` you will find out that the hash changed.

---

On #1253, as far as we could tell, asserting in `Inode.find` that every output inode respects the `check_stable` invariant, breaks the unit tests.
